# **NOMS I COGNOMS:** Junjie Li, Manuel Liu Wang, Leying Wang

## **GRUP DE LABORATORI:** A

# Recomanador Simple

## 1. INFORMACIÓ 

### 1.1. Abans de començar...

**\+ Durant la pràctica, solament es podran fer servir les següents llibreries**:

`Pandas, Numpy, Itertools`

*Nota: A més de les que ja es troben presents en la 1a cel·la i funcions natives de Python*

**\+ No es poden modificar les definicions de les funcions donades, ni canviar els noms de les variables i paràmetres ja donats**

Això no implica però que els hàgiu de fer servir. És a dir, que la funció tingui un paràmetre anomenat `df` no implica que l'hàgiu de fer servir, si no ho trobeu convenient.

**\+ En les funcions, s'especifica que serà i de quin tipus cada un dels paràmetres, cal respectar-ho**

Per exemple (ho posarà en el pydoc de la funció), `df` sempre serà indicatiu del `Pandas.DataFrame` de les dades. Durant la correcció, els paràmetres (i específicament `df`) no contindran les mateixes dades que en aquest notebook, si bé si seran del mateix tipus! Per tant, no us refieu de què tinguin, per exemple, el mateix nombre de files.

### 1.2. Consum general

La base de dades [movielens-1M](http://www.grouplens.org/node/73) conté 1,000,209 puntuacions de 3.900 pel·licules fetes l'any 2000 per 6.040 usuaris anònims del recomanador online [MovieLens](http://www.movielens.org/). 

El consum total de tots els usuaris s'hi pot trobar al document "ratings.dat" el format següent:

    UserID::MovieID::Rating::Timestamp

- **UserIDs** usuaris amb id's entre 1 i 6040 
- **MovieIDs** pelis amb id's entre 1 i 3952
- **Ratings** són les puntuacions en una escala de 5 estrelles.
- **Timestamp** representat en segons

> Cada usuari té com a mínim 20 interaccions consumides.

### 1.3. Usuaris



Al fitxer "users.dat" hi trobem la informació referent a cadascun dels usuaris en el següent format:

        UserID::Gender::Age::Occupation::Zip-code

- **Gender** ve donat per "M" per home i "F" per dona.
- **Age** està representada de la següent forma:

	*  1:  "Under 18"
	* 18:  "18-24"
	* 25:  "25-34"
	* 35:  "35-44"
	* 45:  "45-49"
	* 50:  "50-55"
	* 56:  "56+"

- **Occupation** es tria entre les següents opcions:

	*  0:  "other" or not specified
	*  1:  "academic/educator"
	*  2:  "artist"
	*  3:  "clerical/admin"
	*  4:  "college/grad student"
	*  5:  "customer service"
	*  6:  "doctor/health care"
	*  7:  "executive/managerial"
	*  8:  "farmer"
	*  9:  "homemaker"
	* 10:  "K-12 student"
	* 11:  "lawyer"
	* 12:  "programmer"
	* 13:  "retired"
	* 14:  "sales/marketing"
	* 15:  "scientist"
	* 16:  "self-employed"
	* 17:  "technician/engineer"
	* 18:  "tradesman/craftsman"
	* 19:  "unemployed"
	* 20:  "writer"

> Els usuaris han donat la informació voluntariament. Així doncs, alguns usuaris poden no tenir informació.


### 1.4. Películes



Al fitxer "movies.dat" hi trobem la informació referent a cadascuna de les películes en el següent format:

        MovieID::Title::Genres

- **Titles** són identics als titols de la base de dades IMDB, incloent l'any de llançament.
- **Genres** de les películes estan separats i seleccionats d'entre els següents:

	* Action
	* Adventure
	* Animation
	* Children's
	* Comedy
	* Crime
	* Documentary
	* Drama
	* Fantasy
	* Film-Noir
	* Horror
	* Musical
	* Mystery
	* Romance
	* Sci-Fi
	* Thriller
	* War
	* Western

> Algunes películes poden tenir el ID malament degut a duplicats accidentals.
>
> Les películes s'han entrat manualment, així que altres inconsistencies poden existir. 

## 2. COMENCEM: explorem les dades!!

### 2.1. Descarregar i llegir dades

+ Baixa't els fitxers que composen la base de dades i els còpies al teu directori de treball. 

+ Llegeix les tres taules de la base de dades en tres DataFrames de pandas amb aquest codi:

In [335]:
!wget https://files.grouplens.org/datasets/movielens/ml-1m.zip
!unzip ml-1m.zip

"wget" no se reconoce como un comando interno o externo,
programa o archivo por lotes ejecutable.
unzip:  cannot find or open ml-1m.zip, ml-1m.zip.zip or ml-1m.zip.ZIP.


In [336]:
import pandas as pd
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_table('ml-1m/users.dat', sep='::', header=None, names=unames, engine='python')
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table('ml-1m/ratings.dat', sep='::', header=None, names=rnames, engine='python')
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_table('ml-1m/movies.dat', sep='::', header=None, names=mnames, engine='python', encoding='latin-1')


### 2.2. Inspecció de les taules

In [337]:
users[:10]

user_id gender  age  occupation    zip
0        1      F    1          10  48067
1        2      M   56          16  70072
2        3      M   25          15  55117
3        4      M   45           7  02460
4        5      M   25          20  55455
5        6      F   50           9  55117
6        7      M   35           1  06810
7        8      M   25          12  11413
8        9      M   25          17  61614
9       10      F   35           1  95370

In [338]:
users[-10:]

user_id gender  age  occupation    zip
6030     6031      F   18           0  45123
6031     6032      M   45           7  55108
6032     6033      M   50          13  78232
6033     6034      M   25          14  94117
6034     6035      F   25           1  78734
6035     6036      F   25          15  32603
6036     6037      F   45           1  76006
6037     6038      F   56           1  14706
6038     6039      F   45           0  01060
6039     6040      M   25           6  11106

In [339]:
ratings[-10:]

user_id  movie_id  rating  timestamp
1000199     6040      2022       5  956716207
1000200     6040      2028       5  956704519
1000201     6040      1080       4  957717322
1000202     6040      1089       4  956704996
1000203     6040      1090       3  956715518
1000204     6040      1091       1  956716541
1000205     6040      1094       5  956704887
1000206     6040       562       5  956704746
1000207     6040      1096       4  956715648
1000208     6040      1097       4  956715569

In [340]:
ratings[:10]

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291
5        1      1197       3  978302268
6        1      1287       5  978302039
7        1      2804       5  978300719
8        1       594       4  978302268
9        1       919       4  978301368

In [341]:
ratings.sort_values('movie_id')[:5]


user_id  movie_id  rating  timestamp
427702     2599         1       4  973796689
1966         18         1       4  978154768
683688     4089         1       5  965428947
596207     3626         1       4  966594018
465902     2873         1       5  972784317

In [342]:
movies[:5]

movie_id                               title                        genres
0         1                    Toy Story (1995)   Animation|Children's|Comedy
1         2                      Jumanji (1995)  Adventure|Children's|Fantasy
2         3             Grumpier Old Men (1995)                Comedy|Romance
3         4            Waiting to Exhale (1995)                  Comedy|Drama
4         5  Father of the Bride Part II (1995)                        Comedy

In [343]:
ratings[:5]

user_id  movie_id  rating  timestamp
0        1      1193       5  978300760
1        1       661       3  978302109
2        1       914       3  978301968
3        1      3408       4  978300275
4        1      2355       5  978824291

### **Exemple:** Fent càlculs sobre DataFrames.

Suposa que volem calcular les **puntuacions mitjanes d'una pel·licula per sexe o edat**. 


El primer pas a obtenir una única estructura que contingui tota la informació. Per fer-ho podem usar la funció ``merge`` de pandas. Aquesta funció infereix automàticament quines columnes ha d'usar per fer el ``merge`` basant-se en els noms que fan intersecció:

In [344]:
data = pd.merge(pd.merge(ratings, users), movies)

# Visualitzem la taula ordenada per identificar d'usuari
data.sort_values(by='user_id')[:10]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0            1      1193       5  978300760      F    1          10  48067   
28501        1        48       5  978824351      F    1          10  48067   
13819        1       938       4  978301752      F    1          10  48067   
51327        1      1207       4  978300719      F    1          10  48067   
31152        1      1721       4  978300055      F    1          10  48067   
37916        1      2762       4  978302091      F    1          10  48067   
18472        1      2687       3  978824268      F    1          10  48067   
45685        1      2692       4  978301570      F    1          10  48067   
22832        1       720       3  978300760      F    1          10  48067   
32771        1       745       3  978824268      F    1          10  48067   

                                                   title  \
0                 One Flew Over the Cuckoo's Nest (1975)   
28501                                  Pocahontas (1995)   
13819                                        Gigi (1958)   
51327                       To Kill a Mockingbird (1962)   
31152                                     Titanic (1997)   
37916                            Sixth Sense, The (1999)   
18472                                      Tarzan (1999)   
45685                   Run Lola Run (Lola rennt) (1998)   
22832  Wallace & Gromit: The Best of Aardman Animatio...   
32771                              Close Shave, A (1995)   

                                     genres  
0                                     Drama  
28501  Animation|Children's|Musical|Romance  
13819                               Musical  
51327                                 Drama  
31152                         Drama|Romance  
37916                              Thriller  
18472                  Animation|Children's  
45685                  Action|Crime|Romance  
22832                             Animation  
32771             Animation|Comedy|Thriller

La funció ``iloc`` ens permet obtenir un subconjunt de files i/o columnes:

In [345]:
data.iloc[3:5]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
3       15      1193       4  978199279      M   25           7  22903   
4       17      1193       5  978158471      M   50           1  95350   

                                    title genres  
3  One Flew Over the Cuckoo's Nest (1975)  Drama  
4  One Flew Over the Cuckoo's Nest (1975)  Drama

Els índexs Boolans ens permeten seleccionar una part de la taula que compleix una condició.

In [346]:
# comptem quin tant per cent de ratings estan fets per una dona
print(data[data['gender']=='F']['rating'].count())
print(float(data['rating'].count()))
print(data[data['gender']=='F']['rating'].count()/float(data['rating'].count()), '%')

246440
1000209.0
0.24638850480249627 %


Per obtenir les **puntuacions mitjanes de cada pel·licula agrupada per edat** podem usar el mètode ``pivot_table`` que és una forma de "canviar" la forma de la taula especificant quin valor agregat (mitjançant una funció predefinida) hi volem en funció dels valors de dues columnes:

In [347]:
mean_ratings = data.pivot_table(values= 'rating', index='title', columns='age', aggfunc='mean')
mean_ratings[:10]

age                                      1         18        25        35  \
title                                                                       
$1,000,000 Duck (1971)                  NaN  3.000000  3.090909  3.133333   
'Night Mother (1986)               2.000000  4.666667  3.423077  2.904762   
'Til There Was You (1997)          3.500000  2.500000  2.666667  2.900000   
'burbs, The (1989)                 4.500000  3.244444  2.652174  2.818182   
...And Justice for All (1979)      3.000000  3.428571  3.724138  3.657143   
1-900 (1994)                            NaN       NaN  2.000000       NaN   
10 Things I Hate About You (1999)  3.745455  3.415020  3.432950  3.102941   
101 Dalmatians (1961)              3.514286  3.295082  3.613757  3.826087   
101 Dalmatians (1996)              3.088235  2.467742  2.928571  3.279570   
12 Angry Men (1957)                4.176471  4.032609  4.408654  4.358333   

age                                      45        50        56  
title                                                            
$1,000,000 Duck (1971)             2.000000  2.750000       NaN  
'Night Mother (1986)               3.833333  3.555556  4.333333  
'Til There Was You (1997)          2.333333  2.500000  2.666667  
'burbs, The (1989)                 2.545455  3.208333  2.666667  
...And Justice for All (1979)      4.100000  3.551724  3.928571  
1-900 (1994)                            NaN       NaN  3.000000  
10 Things I Hate About You (1999)  3.258065  3.629630  4.000000  
101 Dalmatians (1961)              3.976744  3.650000  3.190476  
101 Dalmatians (1996)              3.482759  3.400000  3.555556  
12 Angry Men (1957)                4.274194  4.287879  4.235294

Per obtenir les **puntuacions mitjanes de cada pel·licula agrupada per sexe**:

In [348]:
mean_ratings = data.pivot_table('rating', index='title',columns='gender', aggfunc='mean')
mean_ratings[:10]

gender                                    F         M
title                                                
$1,000,000 Duck (1971)             3.375000  2.761905
'Night Mother (1986)               3.388889  3.352941
'Til There Was You (1997)          2.675676  2.733333
'burbs, The (1989)                 2.793478  2.962085
...And Justice for All (1979)      3.828571  3.689024
1-900 (1994)                       2.000000  3.000000
10 Things I Hate About You (1999)  3.646552  3.311966
101 Dalmatians (1961)              3.791444  3.500000
101 Dalmatians (1996)              3.240000  2.911215
12 Angry Men (1957)                4.184397  4.328421

Si volgéssim fer càlculs només sobre les pel·licules que han rebut **al menys** 250 puntuacions, primer hem de construir una taula amb el nombre d'avaluacions de cada títol. Per fer-ho, agruparem les dades per títol (amb el mètode ``groupby``) i usarem ``size()`` el nombre.

El mètode ``groupby`` implenta un o més d'aquests processos:

+ Dividir les dades segons algun criteri.
+ Aplicar una funció a cada grup.
+ Combinar els resultats en una estructura de dades.

In [349]:
ratings_by_title = data.groupby('title').size()
print(ratings_by_title)

title
$1,000,000 Duck (1971)                         37
'Night Mother (1986)                           70
'Til There Was You (1997)                      52
'burbs, The (1989)                            303
...And Justice for All (1979)                 199
                                             ... 
Zed & Two Noughts, A (1985)                    29
Zero Effect (1998)                            301
Zero Kelvin (Kjærlighetens kjøtere) (1995)      2
Zeus and Roxanne (1997)                        23
eXistenZ (1999)                               410
Length: 3706, dtype: int64


Llavors podem crear un índex amb els títols amb més de 250 avaluacions.

In [350]:
active_titles = ratings_by_title.index[ratings_by_title >= 250]
active_titles

Index([''burbs, The (1989)', '10 Things I Hate About You (1999)',
       '101 Dalmatians (1961)', '101 Dalmatians (1996)', '12 Angry Men (1957)',
       '13th Warrior, The (1999)', '2 Days in the Valley (1996)',
       '20,000 Leagues Under the Sea (1954)', '2001: A Space Odyssey (1968)',
       '2010 (1984)',
       ...
       'X-Men (2000)', 'Year of Living Dangerously (1982)',
       'Yellow Submarine (1968)', 'You've Got Mail (1998)',
       'Young Frankenstein (1974)', 'Young Guns (1988)',
       'Young Guns II (1990)', 'Young Sherlock Holmes (1985)',
       'Zero Effect (1998)', 'eXistenZ (1999)'],
      dtype='object', name='title', length=1216)

L'índex de títols que reben al menys 250 puntuacions es pot fer servir per seleccionar les files de ``mean_ratings``: 

In [351]:
mean_ratings = mean_ratings.loc[active_titles]
mean_ratings

gender                                    F         M
title                                                
'burbs, The (1989)                 2.793478  2.962085
10 Things I Hate About You (1999)  3.646552  3.311966
101 Dalmatians (1961)              3.791444  3.500000
101 Dalmatians (1996)              3.240000  2.911215
12 Angry Men (1957)                4.184397  4.328421
...                                     ...       ...
Young Guns (1988)                  3.371795  3.425620
Young Guns II (1990)               2.934783  2.904025
Young Sherlock Holmes (1985)       3.514706  3.363344
Zero Effect (1998)                 3.864407  3.723140
eXistenZ (1999)                    3.098592  3.289086

[1216 rows x 2 columns]

Per veure els films més valorats per les dones, podem ordenar per la columna F de forma descendent:

In [352]:
top_female_ratings = mean_ratings.sort_values(by='F', ascending=False)
top_female_ratings[:10]

gender                                                     F         M
title                                                                 
Close Shave, A (1995)                               4.644444  4.473795
Wrong Trousers, The (1993)                          4.588235  4.478261
Sunset Blvd. (a.k.a. Sunset Boulevard) (1950)       4.572650  4.464589
Wallace & Gromit: The Best of Aardman Animation...  4.563107  4.385075
Schindler's List (1993)                             4.562602  4.491415
Shawshank Redemption, The (1994)                    4.539075  4.560625
Grand Day Out, A (1992)                             4.537879  4.293255
To Kill a Mockingbird (1962)                        4.536667  4.372611
Creature Comforts (1990)                            4.513889  4.272277
Usual Suspects, The (1995)                          4.513317  4.518248

Suposem ara que volem les pel·licules que estan valorades de forma més diferent entre homes i dones. Una forma d'obtenir-ho és afegir una columna a ``mean_ratings`` que contingui la diferència en mitjana i llavors ordenar:

In [353]:
mean_ratings['diff'] = mean_ratings['M'] - mean_ratings['F']
mean_ratings

gender                                    F         M      diff
title                                                          
'burbs, The (1989)                 2.793478  2.962085  0.168607
10 Things I Hate About You (1999)  3.646552  3.311966 -0.334586
101 Dalmatians (1961)              3.791444  3.500000 -0.291444
101 Dalmatians (1996)              3.240000  2.911215 -0.328785
12 Angry Men (1957)                4.184397  4.328421  0.144024
...                                     ...       ...       ...
Young Guns (1988)                  3.371795  3.425620  0.053825
Young Guns II (1990)               2.934783  2.904025 -0.030758
Young Sherlock Holmes (1985)       3.514706  3.363344 -0.151362
Zero Effect (1998)                 3.864407  3.723140 -0.141266
eXistenZ (1999)                    3.098592  3.289086  0.190494

[1216 rows x 3 columns]

Ordenant per ``diff`` ens dóna les pel·licules ben valorades per les dones que presenten més diferència entre homes i dones:

In [354]:
sorted_by_diff = mean_ratings.sort_values(by='diff')
sorted_by_diff[:15]

gender                                        F         M      diff
title                                                              
Dirty Dancing (1987)                   3.790378  2.959596 -0.830782
Jumpin' Jack Flash (1986)              3.254717  2.578358 -0.676359
Grease (1978)                          3.975265  3.367041 -0.608224
Little Women (1994)                    3.870588  3.321739 -0.548849
Steel Magnolias (1989)                 3.901734  3.365957 -0.535777
Anastasia (1997)                       3.800000  3.281609 -0.518391
Rocky Horror Picture Show, The (1975)  3.673016  3.160131 -0.512885
Color Purple, The (1985)               4.158192  3.659341 -0.498851
Age of Innocence, The (1993)           3.827068  3.339506 -0.487561
Free Willy (1993)                      2.921348  2.438776 -0.482573
French Kiss (1995)                     3.535714  3.056962 -0.478752
Little Shop of Horrors, The (1960)     3.650000  3.179688 -0.470312
Guys and Dolls (1955)                  4.051724  3.583333 -0.468391
Mary Poppins (1964)                    4.197740  3.730594 -0.467147
Patch Adams (1998)                     3.473282  3.008746 -0.464536

Invertint l'ordre de les files i fent un ``slicing`` de les 15 files superiors obtenim les pel·licules ben valorades pels homes que no han agradat a les dones: 

In [355]:
sorted_by_diff[::-1][:15]

gender                                         F         M      diff
title                                                               
Good, The Bad and The Ugly, The (1966)  3.494949  4.221300  0.726351
Kentucky Fried Movie, The (1977)        2.878788  3.555147  0.676359
Dumb & Dumber (1994)                    2.697987  3.336595  0.638608
Longest Day, The (1962)                 3.411765  4.031447  0.619682
Cable Guy, The (1996)                   2.250000  2.863787  0.613787
Evil Dead II (Dead By Dawn) (1987)      3.297297  3.909283  0.611985
Hidden, The (1987)                      3.137931  3.745098  0.607167
Rocky III (1982)                        2.361702  2.943503  0.581801
Caddyshack (1980)                       3.396135  3.969737  0.573602
For a Few Dollars More (1965)           3.409091  3.953795  0.544704
Porky's (1981)                          2.296875  2.836364  0.539489
Animal House (1978)                     3.628906  4.167192  0.538286
Exorcist, The (1973)                    3.537634  4.067239  0.529605
Fright Night (1985)                     2.973684  3.500000  0.526316
Barb Wire (1996)                        1.585366  2.100386  0.515020

Si volguéssim les pel·licules que han generat puntuacions més discordants, independentment del gènere, podem fer servir la variança o la desviació estàndard de les puntuacions: 

In [356]:
# Standard deviation of rating grouped by title
rating_std_by_title = data.groupby('title')['rating'].std()
# Filter down to active_titles
rating_std_by_title = rating_std_by_title.loc[active_titles]
rating_std_by_title.sort_values(ascending=False)[:10]

title
Dumb & Dumber (1994)                     1.321333
Blair Witch Project, The (1999)          1.316368
Natural Born Killers (1994)              1.307198
Tank Girl (1995)                         1.277695
Rocky Horror Picture Show, The (1975)    1.260177
Eyes Wide Shut (1999)                    1.259624
Evita (1996)                             1.253631
Billy Madison (1995)                     1.249970
Fear and Loathing in Las Vegas (1998)    1.246408
Bicentennial Man (1999)                  1.245533
Name: rating, dtype: float64

### Important: Temes de rendiment

In [357]:
%timeit data['title'] 
%timeit data.title 
%timeit data[['title']] 

3.06 µs ± 775 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
5.99 µs ± 916 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
15.7 ms ± 643 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [358]:
type(data[['title']])

pandas.core.frame.DataFrame

In [359]:
type(data.title)

pandas.core.series.Series

## 3. EXERCICIS

### 3.1. PRIMER EXERCICI 

Donada la taula ``data``, calcula la puntuació mitjana de cada usuari. 

In [360]:
data_folder = 'ml-1m'

In [361]:
import pandas as pd
unames = ['user_id', 'gender', 'age', 'occupation', 'zip']
users = pd.read_table(f'{data_folder}/users.dat', sep='::', header=None, names=unames, engine='python')
rnames = ['user_id', 'movie_id', 'rating', 'timestamp']
ratings = pd.read_table(f'{data_folder}/ratings.dat', sep='::', header=None, names=rnames, engine='python')
mnames = ['movie_id', 'title', 'genres']
movies = pd.read_table(f'{data_folder}/movies.dat', sep='::', header=None, names=mnames, engine='python',encoding='latin-1')

data = pd.merge(pd.merge(ratings, users), movies)
data


user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0              1      1193       5  978300760      F    1          10  48067   
1              2      1193       5  978298413      M   56          16  70072   
2             12      1193       4  978220179      M   25          12  32793   
3             15      1193       4  978199279      M   25           7  22903   
4             17      1193       5  978158471      M   50           1  95350   
...          ...       ...     ...        ...    ...  ...         ...    ...   
1000204     5949      2198       5  958846401      M   18          17  47901   
1000205     5675      2703       3  976029116      M   35          14  30030   
1000206     5780      2845       1  958153068      M   18          17  92886   
1000207     5851      3607       5  957756608      F   18          20  55410   
1000208     5938      2909       4  957273353      M   25           1  35401   

                                               title                genres  
0             One Flew Over the Cuckoo's Nest (1975)                 Drama  
1             One Flew Over the Cuckoo's Nest (1975)                 Drama  
2             One Flew Over the Cuckoo's Nest (1975)                 Drama  
3             One Flew Over the Cuckoo's Nest (1975)                 Drama  
4             One Flew Over the Cuckoo's Nest (1975)                 Drama  
...                                              ...                   ...  
1000204                           Modulations (1998)           Documentary  
1000205                        Broken Vessels (1998)                 Drama  
1000206                            White Boys (1999)                 Drama  
1000207                     One Little Indian (1973)  Comedy|Drama|Western  
1000208  Five Wives, Three Secretaries and Me (1998)           Documentary  

[1000209 rows x 10 columns]

In [362]:
# la vostra solució aquí
mean = data[['user_id','rating']].groupby('user_id').mean()
mean

rating
user_id          
1        4.188679
2        3.713178
3        3.901961
4        4.190476
5        3.146465
...           ...
6036     3.302928
6037     3.717822
6038     3.800000
6039     3.878049
6040     3.577713

[6040 rows x 1 columns]

+ Quina és la pel·lícula més ben puntuada (en mitja) pels usuaris? (Guarda aquest valor en una variable ``string``). 

In [363]:
# la vostra solució aquí
best_movie = data[['rating','title']].groupby(['title']).mean().sort_values(by='rating', ascending=False).index[0]
best_movie, type(best_movie)

('Ulysses (Ulisse) (1954)', str)

### 3.2. SEGON EXERCICI

Defineix una funció anomenada ``top_movie`` que donat un usuari ens retorni quina és la pel·lícula millor puntuada.<br> 


In [364]:
# la vostra solució aquí
def top_movie(dataFrame,usr):
    '''
    retorni la pel·lícula millor puntuada per user determinat

    :param df --> df: DataFrame original 
    :parm usr --> int: id de user
    :return --> str: title de movie
    '''
    return dataFrame[dataFrame['user_id'] == usr].sort_values(by='rating',ascending=False)["title"][0]

print(top_movie(data,1))


One Flew Over the Cuckoo's Nest (1975)


### 3.3. TERCER EXERCICI

Construeix una funció que donat un dataframe et retorni el valor que cada usuari li ha donat a una peli. Això ho farem creant un dataframe on les columnes són els `movie_id`, les files `user_id` i els valors siguin el rating donat.

In [365]:
def build_counts_table(df):
    """
    Retorna un dataframe on les columnes són els `movie_id`, les files `user_id` i els valors
    el nombre de vegades que un usuari ha vist una peli d'un `movie_id`
    
    :param df: DataFrame original 
    :return: DataFrame descrit adalt
    """
    
    # la vostra solució aquí
    return data.pivot_table(values='rating', index='user_id', columns='movie_id')

In [366]:
df_counts = build_counts_table(data)
df_counts

movie_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                               ...   
1          5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5          NaN   NaN   NaN   NaN   NaN   2.0   NaN   NaN   NaN   NaN  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
6036       NaN   NaN   NaN   2.0   NaN   3.0   NaN   NaN   NaN   NaN  ...   
6037       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6038       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6039       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6040       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   

movie_id  3943  3944  3945  3946  3947  3948  3949  3950  3951  3952  
user_id                                                               
1          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
5          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
6036       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
6037       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
6038       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
6039       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
6040       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[6040 rows x 3706 columns]

També programarem una funció que donada la taula anterior i dos id's (usuari i peli), extregui el valor donat:

In [367]:
def get_count(df, user_id, movie_id):
    """
    Retorna el nombre de vegades que un usuari ha comprat en un `movie_id`
    
    :param df: DataFrame retornat per `build_counts_table`
    :param user_id: ID de l'usuari
    :param movie_id: ID de la peli
    :return: Enter amb el nombre de vegades que ha comprat
    """
    
    # la vostra solució aquí
    return df.values[user_id, movie_id]

get_count(df_counts, 6035, 3)

2.0

### 3.4. QUART EXERCICI

In [368]:
data.nunique()

user_id         6040
movie_id        3706
rating             5
timestamp     458455
gender             2
age                7
occupation        21
zip             3439
title           3706
genres           301
dtype: int64

Si observem el nombre total d'usuaris unics i de pelicules úniques, podem observar que els id's dels usuaris van de 1 a 6040. Normalment hauriem d'intentar que comencessin al nombre 0, anant de 0 a 6039. 

**Què passa amb els indexos de les pelis??**

> ANSWER
>> Si llistem la taula data segons l'ordre dels ids de les pelis, podrem observar que l'id més gran que hi ha és el 3952, però amb la funció nunique només hem detectat 3706 ids, això es deu al fet que la funció ens retorna el nombre de pel·lícules diferents disponibles, que no és necessari que coincideixin aquests valors. Aquesta funció només ens dona la informació que hi ha 3706 pel·lícules diferents.

Usant la funció **pd.Categorical(*).codes**, re-indexa els id's dels usuaris i de les pelis perquè vagin de 0 a 6039 i de 0 a 3705 respectivament:

In [369]:
# la vostra solució aquí
data['user_id'] = pd.Categorical(data['user_id']).codes
data['movie_id'] = pd.Categorical(data['movie_id']).codes
data.sort_values(by= 'user_id')

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0             0      1104       5  978300760      F    1          10  48067   
28501         0        47       5  978824351      F    1          10  48067   
13819         0       877       4  978301752      F    1          10  48067   
51327         0      1117       4  978300719      F    1          10  48067   
31152         0      1574       4  978300055      F    1          10  48067   
...         ...       ...     ...        ...    ...  ...         ...    ...   
578459     6039      2439       2  956716343      M   25           6  11106   
338950     6039      1767       4  997454036      M   25           6  11106   
464888     6039       843       4  956716845      M   25           6  11106   
632232     6039      2462       4  957717463      M   25           6  11106   
903432     6039      1085       2  956716317      M   25           6  11106   

                                                  title  \
0                One Flew Over the Cuckoo's Nest (1975)   
28501                                 Pocahontas (1995)   
13819                                       Gigi (1958)   
51327                      To Kill a Mockingbird (1962)   
31152                                    Titanic (1997)   
...                                                 ...   
578459                               Superman II (1980)   
338950                           West Side Story (1961)   
464888                               Rear Window (1954)   
632232            Invasion of the Body Snatchers (1956)   
903432  Cook the Thief His Wife & Her Lover, The (1989)   

                                      genres  
0                                      Drama  
28501   Animation|Children's|Musical|Romance  
13819                                Musical  
51327                                  Drama  
31152                          Drama|Romance  
...                                      ...  
578459               Action|Adventure|Sci-Fi  
338950                       Musical|Romance  
464888                      Mystery|Thriller  
632232                         Horror|Sci-Fi  
903432                                 Drama  

[1000209 rows x 10 columns]

Per comprovar que tot sigui correcte i guardar correctament la taula **df_counts**, torna a calcular-la:

In [370]:
df_counts = build_counts_table(data)
df_counts[:10]

movie_id  0     1     2     3     4     5     6     7     8     9     ...  \
user_id                                                               ...   
0          5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
1          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4          NaN   NaN   NaN   NaN   NaN   2.0   NaN   NaN   NaN   NaN  ...   
5          4.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6          NaN   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   NaN  ...   
7          4.0   NaN   NaN   3.0   NaN   NaN   NaN   NaN   NaN   NaN  ...   
8          5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
9          5.0   5.0   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN  ...   

movie_id  3696  3697  3698  3699  3700  3701  3702  3703  3704  3705  
user_id                                                               
0          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
5          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
6          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
7          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
8          NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN   NaN  
9          NaN   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   NaN  

[10 rows x 3706 columns]

### 3.5. CINQUÉ EXERCICI



Construeix una funció <b>distEuclid(x,y)</b> que implementi la distància Euclidiana entre dos vectors usant funcions de pandas. Escriu la funció que calculin la semblança entre dos usuaris segons aquesta estructura:

``def SimEuclid (DataFrame, User1, User2)``
    Calcular els **embeddings** de cada usuari, C1 i C2, amb les puntuacions dels ítems comuns que han puntuat el dos usuaris.
    Si no hi ha puntuacions en comú, retornar 0. Retornar ``1/(1+distEuclid(C1, C2))``

\
Avalueu amb la funció ``%timeit`` quant triguen aquests càlculs per un parell d'usuaris.   

> *Nota: Alguns d'aquests exercicis tenen temps de càlcul de l'ordre de minuts sobre tota la base de dades. Per desenvolupar els algorismes és recomanable treballar amb una versió reduïda de la base de dades.* 

> *Nota: **embedding** fa referencia a un vector de N dimensions que té com a funció representar (en aquest cas) els gustos dels usuaris. També es pot fer un embedding d'items i aleshores representarien el contingut d'aquest o caracteristiques semblants.*

---
#### Mesurament de similituds (petita teoria)

El primer pas per poder recomanar és definir una funció de similitud entre vectors. Siguin $x, y$ vectors, implementa la seva distància euclidiana:

* Distància euclidea (inversa): https://en.wikipedia.org/wiki/Euclidean_distance

$$sim(x, y) = \frac{1}{1 + \sqrt{\sum_i(x_i-y_i)^2}}\in [0, 1]$$

---

Per implementar qualsevol d'aquestes únicament es permet l'ús de:

* `np.sum`
* `np.sqrt`
* `np.power`
* `np.dot`
* `np.linalg.norm`
* `np.mean`

I s'ha de fer **sense bucles**.


In [371]:
import math
import numpy as np
import pandas as pd

def distEuclid(x, y):
    """
    Retorna la distancia euclidiana de dos vectors n-dimensionals.
    
    :param x: Primer vector
    :param y: Segon vector
    :return : Escalar (float) corresponent a la distancia euclidiana
    """
    
    # la vostra solució aquí
    return np.linalg.norm(x - y)
    

def SimEuclid(DataFrame, User1, User2):
    """
    Retorna un score que representa la similitud entre user1 i user2 basada en la distancia euclidiana
    
    :param DataFrame: dataframe que conté totes les dades
    :param User1: id user1
    :param User2: id user2
    :return : Escalar (float) corresponent al score
    """

    # la vostra solució aquí

    # dades on tant user1 com user2 tenen pel·lícules valorades
    data = DataFrame.loc[[User1, User2]].dropna(axis=1)
   
    return 1/(1 + distEuclid(data.loc[User1],data.loc[User2]))

In [372]:
# Execute functions
# print(SimEuclid(data, 1, 5))
SimEuclid(df_counts, 0, 2)

0.23166247903554

### 3.6. SISÉ EXERCICI

Desenvolupa un sistema de recomanació col·laboratiu **basat en usuaris**. 

La funció principal, ``getRecommendationsUser``, ha de tenir com a entrada una taula de puntuacions, un ``user_id``, el tipus de mesura de semblança (Euclidiana) que volem usar, el nombre `m` d'usuaris semblants que volem per fer la recomanació i el nombre ``n`` de recomanacions que volem. Com a sortida ha de donar la llista de les ``n`` millors pel·lícules que li podriem recomanar segons la seva semblança amb altres usuaris.

> *Nota 1: S'ha d'evitar comparar ``user_id`` a ell mateix.*

> *Nota 2: Recordeu que en Python podem passar funcions com a paràmetres d'una funció.*

In [373]:
# getRecommendationsUser(data, 2, 50, 10, SimEuclid)

#### - SECCIÓ 1: 

Computa la score de similitud del usuari desitjat (userID) respecte tots els altres i retorna una llista dels **m** usuaris més propers, que seran els que usarem per fer la recomanació. Normalitzeu els scores de sortida.

In [374]:
def find_similar_users(DataFrame, userID, m_user, simfunction):
    """
    Retorna un diccionari de usuaris similars amb les scores corresponents.
    
    :param DataFrame: dataframe que conté totes les dades
    :param userID: usuari respecte al qual fem la recomanació
    :param m: nombre d'usuaris que volem per fer la recomanació
    :param similarity: mesura de similitud
    :return : diccionary
    """

    DataFrame = build_counts_table(DataFrame)

    result = dict()
    user_IDs = DataFrame.drop(userID).index

    # obtenim les similituds entre l'usuari actual amb tota la resta d'usuaris
    similaritys = list((map(lambda user_id : simfunction(DataFrame, userID, user_id), user_IDs)))

    result = dict(zip(user_IDs, similaritys))
    
    # ordenem el diccionari en ordre descendent
    result = dict(sorted(result.items(), key=lambda item: item[1], reverse = True)[:m_user])

    return result

In [375]:
import datetime
t = datetime.datetime.now()
sim_dict = find_similar_users(data, 2, 10, SimEuclid)
t = datetime.datetime.now()-t
print(str(t))

0:00:18.920296


In [376]:
sim_dict

{15: 1.0,
 60: 1.0,
 115: 1.0,
 152: 1.0,
 157: 1.0,
 163: 1.0,
 170: 1.0,
 278: 1.0,
 303: 1.0,
 314: 1.0}

**CREIEU QUE ES OPTIM EL TEMPS QUE HA TARDAT PER UN USUARI?** RAONA LA RESPOSTA.

> ANSWER
>> El temps que ha tardat en executar per un sol usuari no és òptim, perquè si més endavant haurem de calcular la similitud per 6040 usuaris, el temps que tardarà per executar tots els usuaris serà molt alt.

Per millorar el problema anteorior, construeix una matriu de mida UXU on cada posició *i,j* indica la distància entre l'element *i* i el *j*. Així doncs, si estàs fent un recomanador basat en usuaris, `matriu[2, 3]` contindrà la similitud entre l'usuari 2 i el 3.

In [377]:
df_counts[:200]

movie_id  0     1     2     3     4     5     6     7     8     9     ...  \
user_id                                                               ...   
0          5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
1          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4          NaN   NaN   NaN   NaN   NaN   2.0   NaN   NaN   NaN   NaN  ...   
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
195        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
196        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
197        5.0   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0  ...   
198        NaN   NaN   3.0   1.0   NaN   NaN   NaN   NaN   3.0   4.0  ...   
199        NaN   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...   

movie_id  3696  3697  3698  3699  3700  3701  3702  3703  3704  3705  
user_id                                                               
0          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
...        ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  
195        NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN   NaN  
196        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
197        NaN   NaN   NaN   NaN   NaN   5.0   NaN   NaN   NaN   NaN  
198        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
199        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[200 rows x 3706 columns]

In [378]:
def similarity_matrix(similarity_function, df_counts):
    """
    Retorna una matriu de mida M x M on cada posició 
    indica la similitud entre usuaris (resp. ítems).
    
    :param similarity_function: Funció que calcularà la similitud 
        entre usuaris (resp. ítems)
    :param df_counts: Dataframe que conté el nombre de vegades que 
        un usuari ha comprat en un `aisle_id`
    :return : Matriu numpy de mida M x M amb les similituds.
    """

    matrix = np.zeros((df_counts.shape[0], df_counts.shape[0]))

    # la vostra solució aquí
    if similarity_function == None:
        # traballem amb matrix
        data  = df_counts.to_numpy()

        for i in range((len(data))//2):
            j = len(data)-2-i

            # Obtenim l'índex de columna sense nan
            i_notnan = ~np.isnan(data[i])
            j_notnan = ~np.isnan(data[j])

            # Calcula distEuclid en la matriu 
            matrix[i,i+1:] = np.nansum(np.power((data[i,i_notnan] - data[i+1:,i_notnan]), 2), axis=1)
            matrix[j,j+1:] = np.nansum(np.power((data[j,j_notnan] - data[j+1:,j_notnan]), 2), axis=1)

        # matrix de similiuds és simètrica
        matrix = matrix + matrix.T
        matrix = 1 / (1 + np.sqrt(matrix))

        # la semblança de l'usuari amb ell mateix és 0
        np.fill_diagonal(matrix, 0)
     
    else:
        # traballem un per un
        matrix = np.zeros((df_counts.shape[0], df_counts.shape[0]))
        for i in range(df_counts.shape[0]):
            for j in range(df_counts.shape[0]):
                matrix[i][j] = similarity_function(df_counts, i, j)

    
    return matrix

Per cridar aquesta funció, el primer paràmetre pot ser:

* Amb la funció de la distancia euclidiana que heu programat abans pot trigar ~@1h 30min treballant directament amb valors de numpy, ~@20h a partir de pandas pur.
* Opcionalment (no és obligatori fer-ho) podeu programar una funció que treballi específicament amb matrius (i no vectors). Si ho feu, cal gestionar-ho quan es rep `None`.  (@5s)

In [379]:
t = datetime.datetime.now()
# sim = similarity_matrix(SimEuclid, df_counts)
sim = similarity_matrix(None, df_counts)
t = datetime.datetime.now()-t
print(str(t))

sim


0:01:38.447031


array([[0.        , 0.33333333, 0.23166248, ..., 1.        , 0.2052131 ,
        0.19074357],
       [0.33333333, 0.        , 0.2052131 , ..., 0.30901699, 0.30901699,
        0.09733669],
       [0.23166248, 0.2052131 , 0.        , ..., 0.5       , 0.21712927,
        0.12389934],
       ...,
       [1.        , 0.30901699, 0.5       , ..., 0.        , 0.30901699,
        0.13802626],
       [0.2052131 , 0.30901699, 0.21712927, ..., 0.30901699, 0.        ,
        0.1310058 ],
       [0.19074357, 0.09733669, 0.12389934, ..., 0.13802626, 0.1310058 ,
        0.        ]])

In [380]:
sim.shape

(6040, 6040)

Ara torna a re-fer la funció **find_similar_users** i mira quant triga... 

> Recorda que les scores han d'estar normalitzades!

In [381]:
def find_similar_users(DataFrame, sim_matrix, userID, m_user):
    """
    Retorna un diccionari de usuaris similars amb les scores corresponents.
    
    :param DataFrame: dataframe que conté totes les dades
    :param sim_matrix: la matriu de similituds per buscar millor users
    :param userID: usuari respecte al qual fem la recomanació
    :param m_user: nombre d'usuaris que volem per fer la recomanació
    :return --> dict: diccionary de similituds de user determinat
    """

    # la vostra solució aquí

    # Utilitzem sim_matrix per ordenar-los segons les seves similituds
    similituds = sim_matrix[userID]
    users = list(np.argsort(-similituds))
    w_scores = list(similituds[users])
    simdict =  dict(zip(users, w_scores))

    users = set(users)
    result = dict()
    # Afegeix usuari segons el l'usuari existent a les dades
    for user_id in DataFrame['user_id'].unique():
        if user_id in users:
            result[user_id] = simdict[user_id]
    
    return dict(sorted(result.items(), key=lambda item: item[1], reverse = True)[:m_user])


In [382]:
t = datetime.datetime.now()
sim_dict2 = find_similar_users(data, sim, 2, 10)
t = datetime.datetime.now()-t

print(str(t))
sim_dict2


0:00:00.030373


{314: 1.0,
 687: 1.0,
 783: 1.0,
 840: 1.0,
 1110: 1.0,
 1173: 1.0,
 1225: 1.0,
 1476: 1.0,
 1481: 1.0,
 1618: 1.0}

**Hauria d'haver baixat més de 30 cops el temps anterior!!**

#### - SECCIÓ 2: 
Computa les recomanacions per un usuari amb cadascun dels m usuaris més propers i fes una funció que retorni la **weighted average list** d'aquests per tal d'obtenir la recomanació final. Feu servir la funció anterior que usava la matriu de similituds per anar més ràpid!!


> *Nota: la **weighted average list** es calcularà agregant els n items més puntuats de cadascun dels m users més semblants al usuari donat.

In [383]:
np.seterr(divide='ignore',invalid='ignore')

def getRecommendationsUser(DataFrame, user, sim_matrix, n_item, m_user):
    """
    Retorna un dataframe de pel·licules amb els scores.
    
    :param DataFrame: dataframe que conté totes les dades
    :param user: usuari al qual fem la recomanació
    :param sim_mx: similarity_function
    :param n: nombre de pelis a recomanar
    :param m: nombre d'usuaris semblants a tenir en compte per les recomanacions
    :return : pandas de pelis amb els seus scores predits
    """
    
    # la vostra solució aquí
    sim_dict = find_similar_users(DataFrame, sim_matrix, user, m_user)

    users = list(sim_dict.keys())
    weights = list(sim_dict.values())

    # Dades sobre els usuaris més semblants i de les pel·lícules que no ha vist
    data = build_counts_table(DataFrame)
    matrix = data.loc[users, data.loc[user].isna()]

    # Valorem les pel·licules no vistes en funció de les similituds d'altres usuaris    
    scores = (np.nan_to_num(matrix.T).dot(weights)) / ( (np.nan_to_num(matrix.T) != 0).dot(weights) )
    index = np.argsort(-scores)[:n_item]

    return pd.DataFrame({'movie_id': list(index), 'weights': scores[index]})

In [384]:

t = datetime.datetime.now()
user_prediction = getRecommendationsUser(data, 3, sim, 200, 500)
t = datetime.datetime.now()-t
print(str(t))


0:00:01.857353


In [385]:
user_prediction

movie_id  weights
0        1410      5.0
1        1742      5.0
2        3438      5.0
3         825      5.0
4        1736      5.0
..        ...      ...
195      1875      4.5
196      1302      4.5
197      1993      4.5
198       874      4.5
199      1472      4.5

[200 rows x 2 columns]

#### - SECCIÓ 3

Comprovem que realment l'usuari 3 no tenia recomanació per les pelis a les que estem recomanant:

In [386]:
data

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0              0      1104       5  978300760      F    1          10  48067   
1              1      1104       5  978298413      M   56          16  70072   
2             11      1104       4  978220179      M   25          12  32793   
3             14      1104       4  978199279      M   25           7  22903   
4             16      1104       5  978158471      M   50           1  95350   
...          ...       ...     ...        ...    ...  ...         ...    ...   
1000204     5948      2017       5  958846401      M   18          17  47901   
1000205     5674      2498       3  976029116      M   35          14  30030   
1000206     5779      2638       1  958153068      M   18          17  92886   
1000207     5850      3367       5  957756608      F   18          20  55410   
1000208     5937      2702       4  957273353      M   25           1  35401   

                                               title                genres  
0             One Flew Over the Cuckoo's Nest (1975)                 Drama  
1             One Flew Over the Cuckoo's Nest (1975)                 Drama  
2             One Flew Over the Cuckoo's Nest (1975)                 Drama  
3             One Flew Over the Cuckoo's Nest (1975)                 Drama  
4             One Flew Over the Cuckoo's Nest (1975)                 Drama  
...                                              ...                   ...  
1000204                           Modulations (1998)           Documentary  
1000205                        Broken Vessels (1998)                 Drama  
1000206                            White Boys (1999)                 Drama  
1000207                     One Little Indian (1973)  Comedy|Drama|Western  
1000208  Five Wives, Three Secretaries and Me (1998)           Documentary  

[1000209 rows x 10 columns]

In [387]:
data_user3 = data[data['user_id']==3]
data_user3

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
28884         3      1025       4  978293964      M   45           7  02460   
45686         3      2488       5  978294230      M   45           7  02460   
46759         3       253       5  978294199      M   45           7  02460   
52257         3      1848       5  978294230      M   45           7  02460   
70281         3      3235       5  978294008      M   45           7  02460   
70732         3      1120       3  978293924      M   45           7  02460   
116540        3       466       4  978294008      M   45           7  02460   
128355        3      1106       2  978294199      M   45           7  02460   
133409        3      1108       5  978294199      M   45           7  02460   
149459        3      1774       5  978294282      M   45           7  02460   
164932        3      3186       4  978294260      M   45           7  02460   
213276        3      2743       4  978294282      M   45           7  02460   
213798        3      1124       4  978294260      M   45           7  02460   
215822        3       971       4  978294282      M   45           7  02460   
217488        3      3460       4  978294260      M   45           7  02460   
218554        3      2173       4  978294230      M   45           7  02460   
219310        3      1288       5  978294199      M   45           7  02460   
221007        3      3294       1  978294008      M   45           7  02460   
222304        3      1111       5  978294230      M   45           7  02460   
223126        3      2739       5  978294230      M   45           7  02460   
224023        3      1148       5  978294260      M   45           7  02460   

                                                    title  \
28884                   E.T. the Extra-Terrestrial (1982)   
45686                    Run Lola Run (Lola rennt) (1998)   
46759           Star Wars: Episode IV - A New Hope (1977)   
52257                          Saving Private Ryan (1998)   
70281                                 Hustler, The (1961)   
70732   Star Wars: Episode VI - Return of the Jedi (1983)   
116540                               Jurassic Park (1993)   
128355  Star Wars: Episode V - The Empire Strikes Back...   
133409                     Raiders of the Lost Ark (1981)   
149459                                       Rocky (1976)   
164932                             Thelma & Louise (1991)   
213276                       Fistful of Dollars, A (1964)   
213798                                       Alien (1979)   
215822                                    Die Hard (1988)   
217488                                     Mad Max (1979)   
218554                                   King Kong (1933)   
219310                                        Jaws (1975)   
221007                                    Predator (1987)   
222304             Good, The Bad and The Ugly, The (1966)   
223126                                  Goldfinger (1964)   
224023                             Terminator, The (1984)   

                                     genres  
28884       Children's|Drama|Fantasy|Sci-Fi  
45686                  Action|Crime|Romance  
46759       Action|Adventure|Fantasy|Sci-Fi  
52257                      Action|Drama|War  
70281                                 Drama  
70732   Action|Adventure|Romance|Sci-Fi|War  
116540              Action|Adventure|Sci-Fi  
128355    Action|Adventure|Drama|Sci-Fi|War  
133409                     Action|Adventure  
149459                         Action|Drama  
164932                         Action|Drama  
213276                       Action|Western  
213798        Action|Horror|Sci-Fi|Thriller  
215822                      Action|Thriller  
217488                        Action|Sci-Fi  
218554              Action|Adventure|Horror  
219310                        Action|Horror  
221007               Action|Sci-Fi|Thriller  
222304                       Action|Western  
223126

In [388]:
desired_movie_id_to_check = 0
best_film_to_predict = user_prediction['movie_id'].iloc[desired_movie_id_to_check] 

best_film_to_predict

1410

In [389]:
data_user3[data_user3['movie_id']==best_film_to_predict]

Empty DataFrame
Columns: [user_id, movie_id, rating, timestamp, gender, age, occupation, zip, title, genres]
Index: []

**Com es que ara estem predint un rating per la pelicula "best_film_to_predict"?** Digues quin nombre és la predicció

In [390]:
# la vostra solució aquí
movie = movies[movies['movie_id'] == best_film_to_predict]
movie

# Podem obtenir el nom de la pel·lícula segons en best_film_to_predict 

movie_id                     title        genres
1387      1410  Evening Star, The (1996)  Comedy|Drama

### 3.7. SETÉ EXERCICI


A continuació usarem la metrica Mean Absolute Error (MAE) per evaluar el nostre sistema. Aquesta mètrica ens permetrà mesurar la diferencia entre dues llistes donat un usuari: 

**1.** La llista amb els ratings originals d'un usuari donat

**2.** La llista de les prediccions generades per aquest usuari

#### - SECCIÓ 1: 

Treu el 10% dels usuaris i reserva aquests en una variable anomenada **test_set** i la resta en una variable anomenada **train_set**.

In [391]:
import random
unique_users = data['user_id'].unique()
n_test_users = int(len(unique_users)*0.1)
n_train_users = len(unique_users) - n_test_users

test_usrs = random.sample(list(unique_users), n_test_users)

In [392]:
n_test_users,n_train_users

(604, 5436)

In [393]:
test_set = data.loc[data['user_id'].isin(test_usrs)]
train_set = data.loc[~data['user_id'].isin(test_usrs)]
assert len(test_set) + len(train_set) == len(data)
test_set.sort_values(by= 'user_id')

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0             0      1104       5  978300760      F    1          10  48067   
43703         0      1781       5  978301590      F    1          10  48067   
45033         0      1782       4  978301753      F    1          10  48067   
45685         0      2488       4  978301570      F    1          10  48067   
1725          0       639       3  978302109      F    1          10  48067   
...         ...       ...     ...        ...    ...  ...         ...    ...   
212239     6037      1059       4  956707708      F   56           1  14706   
815636     6037      1094       5  956717204      F   56           1  14706   
446137     6037      1184       3  956707604      F   56           1  14706   
563374     6037      3165       3  956706827      F   56           1  14706   
957512     6037      1256       3  956714725      F   56           1  14706   

                                         title                        genres  
0       One Flew Over the Cuckoo's Nest (1975)                         Drama  
43703                          Rain Man (1988)                         Drama  
45033                Driving Miss Daisy (1989)                         Drama  
45685         Run Lola Run (Lola rennt) (1998)          Action|Crime|Romance  
1725          James and the Giant Peach (1996)  Animation|Children's|Musical  
...                                        ...                           ...  
212239  Monty Python and the Holy Grail (1974)                        Comedy  
815636             English Patient, The (1996)             Drama|Romance|War  
446137                   Cool Hand Luke (1967)                  Comedy|Drama  
563374                Muppet Movie, The (1979)             Children's|Comedy  
957512               Breaking the Waves (1996)                         Drama  

[91429 rows x 10 columns]

Què passarà si calculo la matriu de similitud amb **train_set** i després intento predir pels usuaris de **test_set**??

> Que el resultat obtingut no tindria sentit ja que la matiru de distàncies està calculada per a un conjunt totalment diferent al conjunt que volem fer la predicció, i de fet, no es podrà.

#### - SECCIÓ 2:

Selecciona aproximadament el 80% de les interaccions de cada usuari de test i junta-les al **train_set**. Ara podem evaluar el sistema?


> Com la pràctica és molt llarga hem programat el codi per un usuari donat i vosaltres només heu de crear la funció que, per cada usuari, afagi el 80% de les intraccions i les unifiqui al dataframe de train.

In [394]:
test_set

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0              0      1104       5  978300760      F    1          10  48067   
2             11      1104       4  978220179      M   25          12  32793   
6             18      1104       5  982730936      M    1          10  48073   
26            95      1104       3  977621789      F   25          16  78028   
72           284      1104       4  976579214      F   35           0  94109   
...          ...       ...     ...        ...    ...  ...         ...    ...   
1000182     4873       607       4  962781918      F   25           4  70808   
1000184     5946      1330       4  957190428      F   45          16  97215   
1000186     5943      1688       1  957197520      F   18          10  27606   
1000196     5327      2245       4  960838075      F   25           4  91740   
1000208     5937      2702       4  957273353      M   25           1  35401   

                                               title                 genres  
0             One Flew Over the Cuckoo's Nest (1975)                  Drama  
2             One Flew Over the Cuckoo's Nest (1975)                  Drama  
6             One Flew Over the Cuckoo's Nest (1975)                  Drama  
26            One Flew Over the Cuckoo's Nest (1975)                  Drama  
72            One Flew Over the Cuckoo's Nest (1975)                  Drama  
...                                              ...                    ...  
1000182                         Condition Red (1995)  Action|Drama|Thriller  
1000184                         Stranger, The (1994)                 Action  
1000186                            Truce, The (1996)              Drama|War  
1000196                         Outside Ozona (1998)         Drama|Thriller  
1000208  Five Wives, Three Secretaries and Me (1998)            Documentary  

[91429 rows x 10 columns]

In [395]:
# Agafem el 20% de les pelis que ha consumit cada usuari de test 
groupby_count = test_set.groupby('user_id')['movie_id'].count()*0.2
groupby_count

user_id
0       10.6
11       4.6
18      51.0
33      32.8
65       5.2
        ... 
5994    37.2
6018    17.8
6024    54.2
6031    20.8
6037     4.0
Name: movie_id, Length: 604, dtype: float64

Seleccionem la posició 1 i aquest use_id serà el que usarem pel codi d'exemple (que després haureu de replicar).

In [396]:
groupby_count.reset_index().iloc[1]

user_id     11.0
movie_id     4.6
Name: 1, dtype: float64

In [397]:
n_test_samples = int(groupby_count.reset_index().iloc[1]['movie_id'])
u = groupby_count.reset_index().iloc[1]['user_id']
u

11.0

In [398]:
test_set_user = test_set[test_set['user_id'] == u]
frame_test = test_set_user.sample(n_test_samples)
print("TOTAL SAMPLES OF THE USER: " + str(len(test_set_user)))
print("TOTAL SAMPLES OF THE USER IN TEST SET: " + str(len(frame_test)))


TOTAL SAMPLES OF THE USER: 23
TOTAL SAMPLES OF THE USER IN TEST SET: 4


In [399]:
test_set_user

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
2            11      1104       4  978220179      M   25          12  32793   
8929         11      2599       5  978220237      M   25          12  32793   
11044        11       858       5  978220120      M   25          12  32793   
133412       11      1108       5  978218949      M   25          12  32793   
135926       11       579       5  978220193      M   25          12  32793   
180939       11      1155       3  978220216      M   25          12  32793   
183406       11      1504       3  978218568      M   25          12  32793   
352131       11      1131       5  978218949      M   25          12  32793   
375401       11       107       5  978220179      M   25          12  32793   
380562       11      3041       4  978218916      M   25          12  32793   
416588       11      1142       3  978220120      M   25          12  32793   
424194       11      1160       3  978220237      M   25          12  32793   
482755       11       862       5  978220237      M   25          12  32793   
489027       11       802       5  978218949      M   25          12  32793   
572856       11      3133       3  978220273      M   25          12  32793   
573621       11       766       3  978218949      M   25          12  32793   
573668       11      3651       4  978218949      M   25          12  32793   
574662       11       873       2  978218568      M   25          12  32793   
574966       11      3416       4  978220216      M   25          12  32793   
575035       11      1211       4  978218916      M   25          12  32793   
575345       11       935       4  978218568      M   25          12  32793   
575631       11      2414       1  978218568      M   25          12  32793   
576255       11      3542       3  978218568      M   25          12  32793   

                                         title  \
2       One Flew Over the Cuckoo's Nest (1975)   
8929                 Christmas Story, A (1983)   
11044                 Wizard of Oz, The (1939)   
133412          Raiders of the Lost Ark (1981)   
135926        Silence of the Lambs, The (1991)   
180939                    Graduate, The (1967)   
183406                  Full Monty, The (1997)   
352131          Godfather: Part II, The (1974)   
375401                      Taxi Driver (1976)   
380562     Hard-Boiled (Lashou shentan) (1992)   
416588             Boat, The (Das Boot) (1981)   
424194                        Chinatown (1974)   
482755                     Citizen Kane (1941)   
489027                   Godfather, The (1972)   
572856                Dog Day Afternoon (1975)   
573621                 Larger Than Life (1996)   
573668                    Almost Famous (2000)   
574662              Father of the Bride (1950)   
574966           Quatermass and the Pit (1967)   
575035       Man Who Would Be King, The (1975)   
575345             2 Days in the Valley (1996)   
575631                       Dick Tracy (1990)   
576255                      Scary Movie (2000)   

                                    genres  
2                                    Drama  
8929                          Comedy|Drama  
11044   Adventure|Children's|Drama|Musical  
133412                    Action|Adventure  
135926                      Drama|Thriller  
180939                       Drama|Romance  
183406                              Comedy  
352131                  Action|Crime|Drama  
375401                      Drama|Thriller  
380562                        Action|Crime  
416588                    Action|Drama|War  
424194          Film-Noir|Mystery|Thriller  
482755                               Drama  
489027                  Action|Crime|Drama  
572856                  Comedy|Crime|Drama  
573621                              Comedy  
573668                        Comedy|Drama  
574662                              Comedy  
574966                              Sci-Fi  
575035    

In [400]:
frame_test

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
574662       11       873       2  978218568      M   25          12  32793   
482755       11       862       5  978220237      M   25          12  32793   
573621       11       766       3  978218949      M   25          12  32793   
380562       11      3041       4  978218916      M   25          12  32793   

                                      title        genres  
574662           Father of the Bride (1950)        Comedy  
482755                  Citizen Kane (1941)         Drama  
573621              Larger Than Life (1996)        Comedy  
380562  Hard-Boiled (Lashou shentan) (1992)  Action|Crime

In [401]:
len(test_set_user.index)

23

In [402]:
frame_train = test_set_user[~test_set_user.index.isin(frame_test.index)]
print("TOTAL SAMPLES OF THE USER IN TRAIN SET: " + str(len(frame_train)))

TOTAL SAMPLES OF THE USER IN TRAIN SET: 19


In [403]:
frame_train

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
2            11      1104       4  978220179      M   25          12  32793   
8929         11      2599       5  978220237      M   25          12  32793   
11044        11       858       5  978220120      M   25          12  32793   
133412       11      1108       5  978218949      M   25          12  32793   
135926       11       579       5  978220193      M   25          12  32793   
180939       11      1155       3  978220216      M   25          12  32793   
183406       11      1504       3  978218568      M   25          12  32793   
352131       11      1131       5  978218949      M   25          12  32793   
375401       11       107       5  978220179      M   25          12  32793   
416588       11      1142       3  978220120      M   25          12  32793   
424194       11      1160       3  978220237      M   25          12  32793   
489027       11       802       5  978218949      M   25          12  32793   
572856       11      3133       3  978220273      M   25          12  32793   
573668       11      3651       4  978218949      M   25          12  32793   
574966       11      3416       4  978220216      M   25          12  32793   
575035       11      1211       4  978218916      M   25          12  32793   
575345       11       935       4  978218568      M   25          12  32793   
575631       11      2414       1  978218568      M   25          12  32793   
576255       11      3542       3  978218568      M   25          12  32793   

                                         title  \
2       One Flew Over the Cuckoo's Nest (1975)   
8929                 Christmas Story, A (1983)   
11044                 Wizard of Oz, The (1939)   
133412          Raiders of the Lost Ark (1981)   
135926        Silence of the Lambs, The (1991)   
180939                    Graduate, The (1967)   
183406                  Full Monty, The (1997)   
352131          Godfather: Part II, The (1974)   
375401                      Taxi Driver (1976)   
416588             Boat, The (Das Boot) (1981)   
424194                        Chinatown (1974)   
489027                   Godfather, The (1972)   
572856                Dog Day Afternoon (1975)   
573668                    Almost Famous (2000)   
574966           Quatermass and the Pit (1967)   
575035       Man Who Would Be King, The (1975)   
575345             2 Days in the Valley (1996)   
575631                       Dick Tracy (1990)   
576255                      Scary Movie (2000)   

                                    genres  
2                                    Drama  
8929                          Comedy|Drama  
11044   Adventure|Children's|Drama|Musical  
133412                    Action|Adventure  
135926                      Drama|Thriller  
180939                       Drama|Romance  
183406                              Comedy  
352131                  Action|Crime|Drama  
375401                      Drama|Thriller  
416588                    Action|Drama|War  
424194          Film-Noir|Mystery|Thriller  
489027                  Action|Crime|Drama  
572856                  Comedy|Crime|Drama  
573668                        Comedy|Drama  
574966                              Sci-Fi  
575035                           Adventure  
575345                               Crime  
575631                        Action|Crime  
576255                       Comedy|Horror

In [404]:
assert len(frame_train) + len(frame_test) == len(test_set_user)
len(frame_train), len(frame_test), len(test_set_user)

(19, 4, 23)

In [405]:
train_set

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
1              1      1104       5  978298413      M   56          16  70072   
3             14      1104       4  978199279      M   25           7  22903   
4             16      1104       5  978158471      M   50           1  95350   
5             17      1104       4  978156168      F   18           3  95825   
7             23      1104       5  978136709      F   25           7  10023   
...          ...       ...     ...        ...    ...  ...         ...    ...   
1000203     5555      2017       3  959445515      M   45           6  92103   
1000204     5948      2017       5  958846401      M   18          17  47901   
1000205     5674      2498       3  976029116      M   35          14  30030   
1000206     5779      2638       1  958153068      M   18          17  92886   
1000207     5850      3367       5  957756608      F   18          20  55410   

                                          title                genres  
1        One Flew Over the Cuckoo's Nest (1975)                 Drama  
3        One Flew Over the Cuckoo's Nest (1975)                 Drama  
4        One Flew Over the Cuckoo's Nest (1975)                 Drama  
5        One Flew Over the Cuckoo's Nest (1975)                 Drama  
7        One Flew Over the Cuckoo's Nest (1975)                 Drama  
...                                         ...                   ...  
1000203                      Modulations (1998)           Documentary  
1000204                      Modulations (1998)           Documentary  
1000205                   Broken Vessels (1998)                 Drama  
1000206                       White Boys (1999)                 Drama  
1000207                One Little Indian (1973)  Comedy|Drama|Western  

[908780 rows x 10 columns]

In [406]:
def add_testdata(traindf, test_set):
    """
    Retorna dos dataframes, un de train i l'altre de test
    
    :param traindf: dataframe que conté les dades de train
    :param test_set: dataframe que conté les dades de test

    :return : 
        - :param 1st: dataframe que conté les dades de train juntament amb el 80% de test seleccionat
        - :param 2nd: dataframe que conté les dades de test que queden (20% restant)
    """
    
    # la vostra solució aquí
    
    unique_users = test_set['user_id'].unique()

    # Un dataframe buit per test set
    testdf = test_set.drop(index=test_set.index)

    groupby_count = test_set.groupby('user_id')['movie_id'].count()*0.2
    
    for i in range(len(unique_users)):
        # seleccionar el 20% de cada usuari
        n_test_samples = int(groupby_count.reset_index().iloc[i]['movie_id'])
        test_set_user = test_set[test_set['user_id'] == groupby_count.reset_index().iloc[i]['user_id']]

        # creeu dataframe temporal per a cada user      
        frame_test = test_set_user.sample(n_test_samples)
        frame_train = test_set_user[~test_set_user.index.isin(frame_test.index)]

        # afegir als seus dataframes
        traindf = pd.concat([traindf, frame_train], ignore_index=True)
        testdf = pd.concat([testdf, frame_test], ignore_index=True)



    return traindf, testdf


In [407]:
train, test = add_testdata(train_set, test_set)

In [408]:
test

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
0            0       144       5  978301777      F    1          10  48067   
1            0      1117       4  978300719      F    1          10  48067   
2            0      2128       3  978302205      F    1          10  48067   
3            0      1782       4  978301753      F    1          10  48067   
4            0       708       3  978824268      F    1          10  48067   
...        ...       ...     ...        ...    ...  ...         ...    ...   
18056     6031      1212       4  956718092      M   45           7  55108   
18057     6037      1288       2  956707005      F   56           1  14706   
18058     6037      1133       5  956707734      F   56           1  14706   
18059     6037      1120       4  956706876      F   56           1  14706   
18060     6037      1256       3  956714725      F   56           1  14706   

                                                   title  \
0                                       Apollo 13 (1995)   
1                           To Kill a Mockingbird (1962)   
2                                   Pleasantville (1998)   
3                              Driving Miss Daisy (1989)   
4                                  Close Shave, A (1995)   
...                                                  ...   
18056          Butch Cassidy and the Sundance Kid (1969)   
18057                                        Jaws (1975)   
18058                            Grand Day Out, A (1992)   
18059  Star Wars: Episode VI - Return of the Jedi (1983)   
18060                          Breaking the Waves (1996)   

                                    genres  
0                                    Drama  
1                                    Drama  
2                                   Comedy  
3                                    Drama  
4                Animation|Comedy|Thriller  
...                                    ...  
18056                Action|Comedy|Western  
18057                        Action|Horror  
18058                     Animation|Comedy  
18059  Action|Adventure|Romance|Sci-Fi|War  
18060                                Drama  

[18061 rows x 10 columns]

In [409]:
train[train['user_id'] == 7]

user_id  movie_id  rating  timestamp gender  age  occupation    zip  \
17667         7      2889       4  978230666      M   25          12  11413   
23409         7       513       4  978229857      M   25          12  11413   
28226         7      1574       5  978230800      M   25          12  11413   
30849         7      2969       4  978230852      M   25          12  11413   
33662         7      1658       4  978231592      M   25          12  11413   
...         ...       ...     ...        ...    ...  ...         ...    ...   
352463        7      1536       5  978230356      M   25          12  11413   
353486        7      1541       5  978230649      M   25          12  11413   
353911        7       962       4  978232203      M   25          12  11413   
354220        7      2494       5  978229347      M   25          12  11413   
355454        7       275       3  978231802      M   25          12  11413   

                                       title                         genres  
17667                      Awakenings (1990)                          Drama  
23409                Schindler's List (1993)                      Drama|War  
28226                         Titanic (1997)                  Drama|Romance  
30849               Girl, Interrupted (1999)                          Drama  
33662         Last Days of Disco, The (1998)                          Drama  
...                                      ...                            ...  
352463                  Boogie Nights (1997)                          Drama  
353486             Joy Luck Club, The (1993)                          Drama  
353911  Robin Hood: Prince of Thieves (1991)                          Drama  
354220                  Arachnophobia (1990)  Action|Comedy|Sci-Fi|Thriller  
355454                           Nell (1994)                          Drama  

[139 rows x 10 columns]

In [410]:
train.shape

(982148, 10)

In [411]:
test[test['user_id'] == 7]

Empty DataFrame
Columns: [user_id, movie_id, rating, timestamp, gender, age, occupation, zip, title, genres]
Index: []

In [412]:
test.shape

(18061, 10)

#### -SECCIÓ 3:

Fes una funció usant **np.abs** que serveixi per evaluar el nostre sistema. **(PISTA:)** La funció ha de mesurar la diferencia en mitjana de la llista de ratings originals amb la llista de ratings predida.

In [413]:
'''
getRecommendationsUserForTrainTestPrediction(df_counts, userdata, user, sim_matrix, n_item, m_user):

Aquí fem una petita modificació al metode getRecommendationsUser de dalt, afegim un paràmetre més: <userdata>

No fem cap modificació al contingut, per millorar l'eficiència del nostre
ús de evaluateRecommendations, hem fet uns canvis pel rendiment.

Quan vam utilitzar originalment getRecommendationsUser, realitzàvem build_counts al dataframe a dins del mètode, 
que requeria molt de temps. Abans, només l'utilitzem per a un sol usuari, de manera que no afectava el rendiment, 
però quan vam començar a utilitzar per a un grup d'usuaris, és un gran problema, hem trobat que quan el mateix 
conjunt de dades només necessita realitzar l'operació build_counts una vegada, altres usuaris poden utilitzar 
els mateixos df_counts, cosa que pot millorar molt el rendiment, així que realitzem l'operació build_count
al dataframe al principi de evaluateRecommendations, i després l'utilitzem com a paràmetre a getRecommendation
'''

def getRecommendationsUserForTrainTestPrediction(df_counts, userdata, user, sim_matrix, n_item, m_user):
    """
    Retorna un dataframe de pel·licules amb els scores.
    
    :param df_counts --> df: dataframe que conté totes les dades (count de cada user per cada movie)
    :param userdata --> df: dataframe que conté totes les dades
    :param user --> int: usuari al qual fem la recomanació
    :param sim_mx --> numpy: similarity_function
    :param n --> int: nombre de pelis a recomanar
    :param m --> int: nombre d'usuaris semblants a tenir en compte per les recomanacions
    :return --> df: pandas de pelis amb els seus scores predits
    """
    
    # la vostra solució aquí
    sim_dict = find_similar_users(userdata, sim_matrix, user, m_user)

    users = list(sim_dict.keys())
    weights = list(sim_dict.values())

    # Dades sobre els usuaris més semblants i de les pel·lícules que no ha vist
    matrix = df_counts.loc[users, df_counts.loc[user].isna()]

    # Valorem les pel·licules no vistes en funció de les similituds d'altres usuaris       
    scores = (np.nan_to_num(matrix.T).dot(weights)) / ( (np.nan_to_num(matrix.T) != 0).dot(weights) )
    index = np.argsort(-scores)[:n_item]

    return pd.DataFrame({'movie_id': list(index), 'weights': scores[index]})

In [414]:
from tqdm import tqdm

def evaluateRecommendations(train, test, m_user, n_item, sim_matrix):
    """
    Retorna l'error generat pel model
    
    :param DataFrame: dataframe que conté totes les dades
    :param userID: usuari respecte al qual fem la recomanació
    :param m: nombre d'usuaris que volem per fer la recomanació
    :param n: nombre de pelis a retornar
    :param sim: matriu de similitud
    :return : Escalar (float) corresponent al MAE
    """
    
    # la vostra solució 
    mae = 0

    user_IDs = test['user_id'].unique()
    train_count = build_counts_table(train)
    test_count = build_counts_table(test)
    
    for user_id in tqdm(user_IDs):     
        # calcular la diferencia en nitjana de la llista de ratings train amb la lista de rating test
        train_weight = getRecommendationsUserForTrainTestPrediction(train_count, train, user_id, sim_matrix, n_item, m_user)['weights'].mean()
        test_prediction = getRecommendationsUserForTrainTestPrediction(test_count, test, user_id, sim_matrix, n_item, m_user)['weights'].mean()
        mae += np.abs(train_weight - test_prediction)

    return mae / (len(user_IDs) * n_item)

In [415]:
t = datetime.datetime.now()
mae = evaluateRecommendations(train, test, 100, 100, sim)
t = datetime.datetime.now()-t
print(str(t))

100%|██████████| 604/604 [00:41<00:00, 14.48it/s]

0:00:44.516313


In [416]:
mae

2.5134106235408583e-05

### 3.8. VUITÉ (I ÚTLIM) EXERCICI


**Que surt més a compte, fer un recomanador unic pels dos sexes o un per cada sexe?** Justifica la resposta per escrit i amb el codi necessari.

In [417]:
pd.options.mode.chained_assignment = None 

def autoEvaluation(data, m_user, n_item):

    # generar train_set i test_set
    unique_users = data['user_id'].unique()
    n_test_users = int(len(unique_users)*0.1)
    n_train_users = len(unique_users) - n_test_users

    test_usrs = random.sample(list(unique_users), n_test_users)

    test_set = data.loc[data['user_id'].isin(test_usrs)]
    train_set = data.loc[~data['user_id'].isin(test_usrs)]

    train, test = add_testdata(train_set, test_set)

    # generar df_counts i matrix similituds
    data['user_id'] = pd.Categorical(data['user_id']).codes
    data['movie_id'] = pd.Categorical(data['movie_id']).codes
    data.sort_values(by= 'user_id')
    df_counts = build_counts_table(data)
    sim_matrix = similarity_matrix(None, df_counts)

    # calcular mae de aquest dada
    mae = evaluateRecommendations(train, test, m_user, n_item, sim_matrix)

    return mae

In [418]:
# la vostra solució aquí

data_F = data[data['gender']=='F']
data_M = data[data['gender']=='M']

mae = autoEvaluation(data, 100, 100)
maeF = autoEvaluation(data_F, 100, 100)
maeM = autoEvaluation(data_M, 100, 100)

100%|██████████| 433/433 [00:34<00:00, 12.66it/s]


In [419]:
print(f"{mae}, {maeF}, {maeM}")
print(f"Recomanador unic pels unic sexes FM: {mae}")
print(f"Recomanador unic pels dos sexes   F: {maeF}")
print(f"Recomanador unic pels dos sexes   M: {maeM}")

maeFM = np.abs(maeM-maeF) / 2
print(f"FM Mitjana: {maeFM}")

2.8125379132166653e-05, 1.8810925580962102e-05, 5.6979733182557883e-05
Recomanador unic pels unic sexes FM: 2.8125379132166653e-05
Recomanador unic pels dos sexes   F: 1.8810925580962102e-05
Recomanador unic pels dos sexes   M: 5.6979733182557883e-05
FM Mitjana: 1.908440380079789e-05


> RESPOSTA

> > Veurem com en els test següents els valors que ens donen pels dos tipus de recomanadors el valors són **MAE** i **maeFM**, assumint que com menor sigui aquest valor, millor és la qualitat del nostre sistema d'avaluació, llavors podem testejar i comparar aquests dos valors, a partir dels resultats podem veure que fer un recomanador de **cada sexes** surt més a compte.